core/trino-main/src/main/java/io/trino/execution/buffer/BufferState.java

```diff
@@ -52,7 +52,12 @@
      * assumed that the reader will be cleaned up elsewhere.
      * This is the terminal state.
      */
-    ABORTED(false, false, true);
+    ABORTED(false, false, true),
+
+    /**
+     * Buffer is failed. No more buffers or pages can be added. The task will be failed.
+     */
+    FAILED(false, false, true);
```

Comment:
Do you need to handle it in `BroadcastOutputBuffer.java` 
like we do for `ABORTED`:
```java
        // do not setup the new buffer if we are already aborted
        if (state != ABORTED) {
            // add initial pages
```

Reply:
The difference between `ABORTED` and `FAILED` is that `ABORTED` indicates that the task itself has been aborted or failed and the buffer is marked as `ABORTED` from the outside (by the task itself). `FAILED` indicates that an internal failure happened inside the buffer and must be reported to the task. The `BroadcastOutputBuffer` does not fail internally so we don't have to handle the `FAILED` state.

Reply:
Yeah - it looked like we cannot get FAILED for `BroadcastOutputBuffer`.
Yet then maybe let's do:

```
 // do not setup the new buffer if we are already aborted
        if (state != ABORTED) {
             verify (state != FAILED);
```

simplifies the reading.